### PR TITLE
fix: Prevent infinite recursion when resolving policy parameter when the parameter refs a CFN parameter with the same name

### PR DESCRIPTION
**Pull request by aahung**
…the parameter refs a CFN parameter with the same name

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [x] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/policy_template_processor/template.py [hunk 1]
```diff
@@ -1,10 +1,13 @@
-import copy
+from typing import Any
 
 from samtranslator.intrinsics.resolver import IntrinsicsResolver
 from samtranslator.intrinsics.actions import RefAction
 from samtranslator.policy_template_processor.exceptions import InsufficientParameterValues, InvalidParameterValues
 
 
+POLICY_PARAMETER_DUSAMBIGUATE_PREFIX = "___SAM_POLICY_PARAMETER_"
```
Comment (acristin) [inline samtranslator/policy_template_processor/template.py:8]: DUS -> DIS

File: samtranslator/policy_template_processor/template.py [hunk 2]
```diff
@@ -50,17 +53,53 @@ def to_statement(self, parameter_values):  # type: ignore[no-untyped-def]
...
-        definition_copy = copy.deepcopy(self.definition)
+        definition_copy = self._disambiguate_policy_parameter(self.definition)
 
         return resolver.resolve_parameter_refs(definition_copy)
 
+    @staticmethod
+    def _disambiguate_policy_parameter(policy_definition: Any) -> Any:
+        """
+        Return a deepcopy of policy definition where all parameters are
+        renamed to avoid naming collusion of normal CFN parameters.
```
Comment (acristin) [inline samtranslator/policy_template_processor/template.py:73]: collusion -> collision

Comment (aahung) [reply samtranslator/policy_template_processor/template.py:73]: damn! thank you Andrew!

**Review by ssenchenko — APPROVED** [review]

File: samtranslator/policy_template_processor/template.py (hunk 2 above)
Comment (aahung) [reply samtranslator/policy_template_processor/template.py]: fixed in a commit named "[Fix a type](https://github.com/aws/serverless-application-model/pull/2751/commits/6a4dd081f3e01d1351e8ea4421f131fc1d269d33)"

**Review by acristin — APPROVED** [review]

**Pull request merged**